app/src/main/java/com/futurice/freesound/feature/search/SoundItemViewModel.kt

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2016 Futurice GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.futurice.freesound.feature.search
+
+import com.futurice.freesound.feature.audio.AudioPlayer
+import com.futurice.freesound.feature.audio.PlaybackSource
+import com.futurice.freesound.feature.audio.PlayerState
+import com.futurice.freesound.feature.audio.from
+import com.futurice.freesound.feature.common.Navigator
+import com.futurice.freesound.network.api.FreeSoundApiService
+import com.futurice.freesound.network.api.model.Sound
+import com.futurice.freesound.viewmodel.SimpleViewModel
+import com.google.auto.factory.AutoFactory
+import com.google.auto.factory.Provided
+import io.reactivex.Observable
+import io.reactivex.Single
+import polanski.option.Option
+import java.text.DateFormat
+
+@AutoFactory
+internal class SoundItemViewModel(private val sound: Sound,
+                                  private @Provided val navigator: Navigator,
+                                  private @Provided val audioPlayer: AudioPlayer,
+                                  private @Provided val freeSoundApiService: FreeSoundApiService) : SimpleViewModel() {
+
+    private val thumbnail: String
+        get() = sound.images.medSizeWaveformUrl
+
+    private val currentPercentage: Observable<Option<Int>>
+        get() = audioPlayer.timePositionMsOnceAndStream
+                .map { positionMs -> toPercentage(positionMs, sound.duration) }
+                .map { Option.ofObj(it) }
+
+    fun thumbnailImageUrl(): Single<String> = Single.just(thumbnail)
+
+    fun name(): Single<String> = Single.just(sound.name)
+
+    fun userAvatar(): Single<String> =
+            freeSoundApiService.getUser(sound.username)
+                    .map { (_, _, avatar) -> avatar.medium }
```

Comment:
Is decomposing the user type here actually any more readable than just mapping via:
`map { user -> user.avatar.medium }`

Reply:
If we have to add more properties to the User class, then `.map { (_, _, avatar) -> avatar.medium }` will have to change, which to me, is unnecessary.